### ISSUE TITLE
Breakpoints are not resolved (they appear greyed in the editor) until just before they are executed

### ISSUE THREAD

**Issue by hasimyerlikaya** [Dart-Code/Dart-Code#4734]
**Describe the bug**
I have a Dart API project, and I can't create breakpoints after updating to v3.72.2. There is no issue with v3.70.0.

**Expected behavior**
Be able to create breakpoint

**Screenshots**
<img width="747" alt="image" src="https://github.com/Dart-Code/Dart-Code/assets/7986394/7b66220e-d52a-4c48-9705-dc2809987e18">

 - Operating System and version: macOS Ventura Version 13.5.2 (22G91)
 - VS Code version: 1.82.0
 - Dart extension version: v3.72.2
 - Dart/Flutter SDK version:  Dart 3.1.1 / Flutter 3.13.3
 
 
```
!! PLEASE REVIEW THIS LOG FOR SENSITIVE INFORMATION BEFORE SHARING !!

Dart Code extension: 3.72.2
Flutter extension: 3.72.0 (activated)

App: Visual Studio Code
App Host: desktop
Version: 1.82.0
Platform: mac

Workspace type: Dart
Analyzer type: LSP
Multi-root?: true

Dart SDK:
    Loc: /Users/hasimyerlikaya/flutter/bin/cache/dart-sdk
    Ver: 3.1.1
Flutter SDK:
    Loc: /Users/hasimyerlikaya/flutter
    Ver: 3.13.3

HTTP_PROXY: undefined
NO_PROXY: undefined

Logging Categories:
    General, AnalyzerTiming, CommandProcesses, DartTest, DAP, VmService, DevTools, FlutterDaemon, FlutterRun, FlutterTest, WebDaemon

Mon Sep 11 2023 [03:18:25 GMT+0300 (GMT+03:00)] Log file started
[3:18:28 AM] [DAP] [Info] ==> {"command":"setBreakpoints","arguments":{"source":{"name":"login_service.dart","path":"/lib/api/user/services/login_service.dart"},"lines":[19],"breakpoints":[{"line":19}],"sourceModified":false},"type":"request","seq":20}
[3:18:28 AM] [DAP] [Info] <== {"seq":30,"type":"response","body":{"breakpoints":[{"id":100007,"verified":false}]},"command":"setBreakpoints","request_seq":20,"success":true}
Mon Sep 11 2023 [03:18:33 GMT+0300 (GMT+03:00)] Log file ended
```

**Comment by DanTup** [comment]
@hasimyerlikaya the breakpoints now appear grey while they are "unresolved" (that is, the VM has not confirmed their locations). They should still work fine and turn red once the VM loads the source and verifies the location (this is usually just before the code executes).

Can you confirm that the breakpoints turn red and function once you actually trigger that code?

**Comment by hasimyerlikaya** [comment]
@DanTup Thank you for reply. 

It is working perfectly, but when I see that it is not red, I think it won't work. In the previous version, it was red all the time. Now, I need to trigger it to ensure that there is no problem with debugging. As you know, when we start a Flutter project in release mode, break points won't be red. In my opinion, the break points should always be red in the dart project, like in the previous version. This version makes me feel that there is an error.


https://github.com/Dart-Code/Dart-Code/assets/7986394/ab77be42-5358-45d5-940a-f082e5f301b5

**Comment by DanTup** [comment]
Thanks for the feedback. I understand that this can be confusing when it worked differently for so long - but this new way does also help understand when breakpoints may be moved by the VM (for example if you put a breakpoint in an invalid location, when it turns red it when it will snap to the nearest location).

I think we may be able to add tooltip text to the grey breakpoints, so perhaps it would be clearer if there was a tooltip that said something like "The location of this breakpoint is not yet resolved" so it's clear it's not an error state?

@jacob314 @bkonyi also interested in your thoughts on this. The VM definitely resolves these locations more lazily then I would've expected, so in most cases the breakpoints are essentially "unresolved until this code executes for the first time" which makes the distinction a little less useful. I presume there's nothing we can do to encourage the VM to resolve these more eagerly?

**Comment by hasimyerlikaya** [comment]
@DanTup You are welcome.

Tooltip will help more. At least we can know that there is no problem. My preference would be a solid gray icon. That way it would be clearer that we are on standby.

**Comment by DanTup** [comment]
We can't change the icon color ourselves, that's handled by VS Code. However I've added some comments to https://github.com/microsoft/debug-adapter-protocol/issues/425 (where I've asked about the semantics of "unverified" breakpoints before) to see whether maybe the debug protocol could support separate states for "failed to verify" and "not yet verified".

**Comment by snipd-min** [comment]
Hi @DanTup, somehow this behaviour feels really awkward for me... It's probably because I'm very used to having the breakpoint as "red" all the time as it's much easier to see where they are and only showing the red breakpoint when a loaded location is invalid (e.g. source file changed and the loaded location weren't in sync) feels more natural.

I totally get that this is the "intended" behaviour as the locations are lazily loaded but in practice it feels quite awkward as the sources are not dynamically loaded, I felt like it's unnecessary to show it as unverified unless the source file was modified and the changes are not yet synced to the dart server... 

This is especially when "launching a debug session", as all files will have to be saved, won't all breakpoints be in theory "verified" even though they are not loaded?

**Comment by DanTup** [comment]
Yep, I understand - what's why I brought this up in the DAP issue above.

I don't think just reverting this behaviour is good, because it loses something - right now the greyed out breakpoints indicate their location is not resolved. That is, you cannot trust they will be hit in this location when they are hit. I think this is useful information (because I've had questions about breakpoints not being hit when they've been added to invalid locations before).

I've pinged on the issue above again and will see if any progress can be made there.

**Comment by dedvalson** [comment]
What I am seeing does not match the descriptions given above. Please tell me if this is the intended behavior:

1. Create new application using Flutter: new project.
2. Place some breakpoints. They are now orange dots.
3. Press F5 to debug using VSCode on Windows and an Android Simulator.
4. Dots are now shown as white hollow circles. But they work properly. 
5. Breakpoints never turn orange again until I stop debugging. Even if I hit the breakpoint.

**Comment by dedvalson** [comment]
I don't think that what I am seeing matches the expected behavior above. Here is what I did:

1. Create new application using Flutter: new project.
2. Put a breakpoint in it. It shows up as orange. 
3. Start debugging using F5 (Windows and Android Emulator).
4. The breakpoint changes to a white circle, then goes back to an orange dot once the app is running.
5. Hit the breakpoint, it works properly. 
6. Add a second breakpoint. **Both breakpoints now turn into white circles.**
7. Now, no matter what I do I never see an orange dot again until I stop debugging.

Any change in breakpoints made after the app starts running causes all breakpoints to change to white circles until debugging ends.

**Comment by bkonyi** [comment]
Sorry for the delayed response, I'm just catching up on emails now.

> @jacob314 @bkonyi also interested in your thoughts on this. The VM definitely resolves these locations more lazily then I would've expected, so in most cases the breakpoints are essentially "unresolved until this code executes for the first time" which makes the distinction a little less useful. I presume there's nothing we can do to encourage the VM to resolve these more eagerly?

I think we might be able to trigger a compilation of unexecuted code when setting a breakpoint, which should result in a resolved breakpoint being returned. I can't say with certainty that we don't trigger a compilation of unexecuted code when setting a breakpoint already, so I'll have to look into this.

**Comment by DanTup** [comment]
@dedvalson 

> I don't think that what I am seeing matches the expected behavior above. Here is what I did:
> ...
> Add a second breakpoint. Both breakpoints now turn into white circles.

Thanks for the clarification. This sounds like another bug https://github.com/Dart-Code/Dart-Code/issues/4469 that was expected to have been fixed. Can you confirm the version of the Flutter SDK and which platform you're seeing this on, and if possible provide exact steps to reproduce (including exactly which line(s) you're adding breakpoints to)?

Thanks!


@bkonyi 

> I think we might be able to trigger a compilation of unexecuted code when setting a breakpoint, which should result in a resolved breakpoint being returned. I can't say with certainty that we don't trigger a compilation of unexecuted code when setting a breakpoint already, so I'll have to look into this.

Ah, that sounds like it would be perfect. If you think it should already be happening and need some steps to repro this issue, let me know.

**Comment by dedvalson** [comment]
Steps to reproduce:

1. Create a new Flutter application.
2. Place a breakpoint on line 67 (      _counter++;)  The breakpoint is red.
3. Press F5 to start debugging. Breakpoint changes to white circle while compiling / launching.
4. Once app is running breakpoint turns red again.
5. Press the "+" floating action button in the app. The breakpoint is hit and stays red.
6. Press F5 again to resume running. 
7. Place a breakpoint on line 79 (return Scaffold). Both breakpoints change to white circles.
8. Press the "+" button again.
9. The line 67 breakpoint is hit. Both breakpoints remain as white circles.
10. Press F5 to continue running.
11. The line 79 breakpoint is hit. Both breakpoints remain as white circles.

I am running DartCode 3.72.2

Here is the output of flutter doctor -v

```
[√] Flutter (Channel stable, 3.13.4, on Microsoft Windows [Version 10.0.19045.3324], locale en-US)
    • Flutter version 3.13.4 on channel stable at c:\Src\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 367f9ea16b (5 days ago), 2023-09-12 23:27:53 -0500
    • Engine revision 9064459a8b
    • Dart version 3.1.2
    • DevTools version 2.25.0

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Don\AppData\Local\Android\sdk
    • Platform android-33, build-tools 30.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Community 2019 16.6.2)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.6.30204.135
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 2022.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)

[√] IntelliJ IDEA Community Edition (version 2018.3)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2018.3.3
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin version 183.5153.38

[√] VS Code (version 1.82.2)
    • VS Code at C:\Users\Don\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.34.0

[√] Connected device (4 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • Windows (desktop)                  • windows       • windows-x64    • Microsoft Windows [Version 10.0.19045.3324]
    • Chrome (web)                       • chrome        • web-javascript • Google Chrome 116.0.5845.180
    • Edge (web)                         • edge          • web-javascript • Microsoft Edge 113.0.1774.42

[√] Network resources
    • All expected network resources are available.

• No issues found!
```

Here is the output from VSCode About

```
Version: 1.82.2 (user setup)
Commit: abd2f3db4bdb28f9e95536dfa84d8479f1eb312d
Date: 2023-09-14T05:55:25.390Z
Electron: 25.8.1
ElectronBuildId: 23779380
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Windows_NT x64 10.0.19045

The Target is an Android Simulator running API 29.

I see exactly the same symptoms when the target is a native Windows application. 
```

**Comment by DanTup** [comment]
@dedvalson thank you for the steps. I was able to reproduce the issue on stable, and it appears fixed on master (and presumably beta).

I believe this was actually fixed by a later change https://github.com/dart-lang/sdk/commit/4d1b41823061b17dea1418e45688d2f8dc8b2c27 which was not included in the current stable release. At the time I made the fix I thought only apps with multiple isolates were affected, but it appears not. The changes made the breakpoint resolution more resilient to timing issues and I suspect that's fixed this too.

I don't know if you're able to easily test on beta/master to confirm, but if you do see it on either of them (or a future stable release), please do file a new issue for it with the steps. Thanks!

**Comment by dedvalson** [comment]
Just to clarify.

Are you referring to the Beta / Master branch of Flutter itself or is there
a Beta / Master branch of Dart Code that is separate?

Don

**Comment by DanTup** [comment]
> Are you referring to the Beta / Master branch of Flutter itself or is there
> a Beta / Master branch of Dart Code that is separate?

I mean of Flutter. There is a pre-release channel for the Dart-Code extension, but the debug adapter code in question ships as part of the Dart/Flutter SDK (in DDS).

**Comment by bkonyi** [comment]
I have a pending [CL](https://dart-review.googlesource.com/c/sdk/+/327360) that should resolve this (no pun intended 😅). Unfortunately, it'll be blocked until https://github.com/dart-lang/sdk/issues/53615 is resolved as there was some bad behavior around setting breakpoints within enhanced enums that just so happened to work correctly when resolving unresolved breakpoints.

**Comment by DanTup** [comment]
Awesome, thanks! 😊

**Comment by DanTup** [comment]
@intonarumori not being red/verified is purely a cosmetic issue and doesn't affect the behaviour.

> Breakpoints don't appear where I add them

This sounds like a different issue to what's described here (and one that is certainly more than cosmetic) - could you please file a new issue with details and a repro so I can take a look? Thanks!

**Comment by intonarumori** [comment]
@DanTup Sure, I will double check my setup and file an issue.

On the cosmetic note:
Most other dev tools like Xcode and Android Studio use filled / unfilled circle to signal if the breakpoint is enabled, not if it's resolved, so I'd argue we are going against the grain here.

Looking at the Android Studio manual it seems they have a checkmark on the breakpoint, that looks like a better way to differentiate between the two and keep everything consistent across different IDEs.

https://developer.android.com/studio/debug

**Comment by DanTup** [comment]
> Sure, I will double check my setup and file an issue.

Thanks!


> Most other dev tools like Xcode and Android Studio use filled / unfilled circle to signal if the breakpoint is enabled, not if it's resolved, so I'd argue we are going against the grain here.

The UI here is controlled by VS Code - the grey circles are how unverified breakpoints are rendered by VS Code. It's a bug that the breakpoints are verifying late (and @bkonyi has started a fix for this), but if the visuals of unverified vs disabled breakpoints seems bad, that would need to be filed https://github.com/microsoft/vscode.

**Comment by bksubhuti** [comment]
I spent a couple of hours trying to resolve this until I figured out that the breakpoints actually work.
I placed an issue here... https://github.com/flutter/flutter/issues/136362#issuecomment-1758966985
I did not get this behavior on my windows machine.  Although I was using a lower version  
now I'm using 3.13.6 on linux mint.  I thought it was a linux or vscode problem because I remember you need to check off some boxes that are not default (not sure why debugging breakpoints are not defualt).. but it still didn't work.

I was directed to this issue.. but I think it is different.  You seem to imply that this is a line that is not executtable.  
I can sort of understand that.
I'm putting breaks on count++; for the counter app and it still does not work (visually red).

**Comment by DanTup** [comment]
@bksubhuti yep, this issue is entirely cosmetic and about breakpoints appearing grey (unverified) until just before they are hit. If you are seeing breakpoints not being hit, that sounds like a different (and more severe) issue. Please file a new issue (here or in the SDK and CC me) with a repro. Thanks!

**Comment by lukehutch** [comment]
> This is definitely a regression, for me it's almost unusable:
> 
> * Breakpoints don't appear where I add them
> * they are not red
> * they work when executing code though

The first 2 points are true for me -- this is extremely unsettling, and I don't think this is good behavior, any way you look at it.

However, the 3rd point doesn't always apply. I don't know how reproducible this is, but just now I had some unconfirmed breakpoints in my code that weren't even hit when the line was executed. That further undermines my confidence in the breakpoint system, beyond the uncertainty introduced by the first two points.

**Comment by DanTup** [comment]
@lukehutch looks like you may have deleted your comment as I was replying - but to re-iterate the above, if you are seeing breakpoints appear in wrong locations or breakpoints not being hit, please do file another issue. This issue is _only_ about breakpoints appearing unverified until they are hit (which is entirely cosmetic). If there are other issues, they need to be debugged and tracked separately, because they won't be fixed by the fix for this issue.

(@bkonyi I saw the issue you were blocked on was fixed.. hopefully this means you can land your fix now too - let me know if you'd like me to do any testing of the changes)

**Comment by lukehutch** [comment]
@DanTup Thanks -- I deleted the comment because the only new point I was making was that breakpoints weren't even being hit during execution, but then I realized it was because I was running in profile mode rather than debug mode.

**Comment by lukehutch** [comment]
Here's an unusual case I found -- I can set a breakpoint successfully on line 136, and the debugger will stop on that line, but if I try to set a breakpoint on any of the nested lines below line 136, the breakpoint shows as unverified, *and* the debugger never stops on the line.

![image](https://github.com/Dart-Code/Dart-Code/assets/811305/7002007a-f664-484b-bafb-5f9771d62e03)

This used to work fine before the breakpoint system was changed; now it's broken. In other words, this is not just a breakpoint rendering issue -- the breakpoints simply don't work anymore under some circumstances.

**Comment by DanTup** [comment]
@lukehutch please file another issue with the details (and ideally a code sample). I understand this might seem related, but the change that's resulting in breakpoints being grey is purely because we return a hard-coded `false` boolean instead of a `true` boolean to VS Code when it's setting breakpoints. Any difference in VM behaviour is another bug that needs investigating and is not a result of this cosmetic issue.

**Comment by bkonyi** [comment]
> One last question on this example though: why would I be able to set a verified breakpoint on the indicated line (136), while other lines below that, in the same function, could not be verified? That does not seem like expected behavior to me, unless a line contains no actual code.

I'm not 100% sure, but my guess is that it's because you're setting a breakpoint in a closure defined within the widget instantiation. This closure is a separate function from `profileEndorsements` so, with the current behaviour, the breakpoint will only be resolved once the closure on line 140 executes.

**Comment by nikolayDemirev** [comment]
I have the same issue. The breakpoints inside a closure never get resolved, even when the code gets executed.

**Comment by DanTup** [comment]
@nikolayDemirev if you're seeing anything other than breakpoints being grey until they are executed, then they turn red (and are correctly hit), then it's likely a different issue. Could you open a new issue with a code sample/steps to reproduce?

This issue is tracking a cosmetic issue where breakpoints are not verified until the code is compiled. The fix is to compile the scripts earlier (when the breakpoints are set), and is unlikely to fix any other issues * (such as breakpoints not being hit, or remaining grey after the code has been executed).

\* unless the issue @bkonyi found while trying to fix this happens to be what you're seeing, in which case it _might_ be fixed incidentally as part of this. I don't know the details of that/those issues though (perhaps Ben can explain), though having another issue with a specific repro would probably be helpful to test and verify without waiting for another release only to find your issues weren't fixed.

**Comment by mrRedSun** [comment]
https://github.com/Dart-Code/Dart-Code/assets/22750848/a5883ce2-ace5-45dd-8e21-1b6b71f8fb81

Flutter 3.22.0, latest VSCode and Dart plugin, break points are unusable and I don't see any pattern to it, sometimes it works, sometimes it doesn't

**Comment by DanTup** [comment]
@mrRedSun it's hard to tell what's happening from the video, but if the issue is more than just that the breakpoints are grey instead of red, please file a new issue with details. There was an issue fixed recently (but is not in the stable release) where breakpoints added after hot restart could jump to the wrong location, but it's not clear if that's the issue here.

**Comment by mrRedSun** [comment]
@DanTup breakpoints disappear when doing a hot reload, but I see from the comments above that a different issue related to breakpoints  has been fixed recently, I'll wait for that to get into stable release and then will retest and will file an issue with repro steps if continues to occur, thanks!

**Comment by DanTup** [comment]
>  I'll wait for that to get into stable release and then will retest

If you're able to provide more details (or do some testing on Flutter `master` to see if it's fixed there), it would certainly be useful to dig into before the next stable release. Flutter releases are approximately quarterly and if it turns out after the next stable release something still isn't working right, it would be longer until a fix does ship (whereas if we can identify any remaining issues soon, they could be fixed in the next release).

**Comment by woutervanwijk** [comment]
Same here, it's really bad

**Comment by lukehutch** [comment]
> Same here, it's really bad

Try `flutter channel master ; flutter upgrade ; flutter pub upgrade` and see if your breakpoint issues are fixed.

**Comment by stan-at-work** [comment]
+1

**Comment by gorkemalgan** [comment]
Same here, master channel is not working as well.

**Comment by stan-at-work** [comment]
> Same here, master channel is not working as well.

Yes the command provided by: @lukehutch does not work

**Comment by DanTup** [comment]
This issue of breakpoints being grey initially is still outstanding. There are some changes being tested that should resolve this (for example https://dart-review.googlesource.com/c/sdk/+/338740) but they have not landed yet.

I will post back here when there is an update (and close it if a fix lands). Please refrain from posting "+1" comments, they generate unnecessary notifications for all those that have comment above. Thanks!

**Comment by DanTup** [comment]
A fix for this has landed in the Dart SDK:

https://github.com/Dart-Code/Dart-Code/assets/1078012/5638a182-fa7a-49da-8805-666561c41a1c

And has also rolled into Flutter `master`:

https://github.com/Dart-Code/Dart-Code/assets/1078012/f1460887-1c3b-4a07-87e0-ecb52710b660

If you see unexpected breakpoint  behaviour when using a version that includes these changes (which is currently only master/bleeding-edge builds, but will eventually be in a beta and then stable release), please file new issues with detailed descriptions of the behaviour you're seeing (and if possible, repro steps/projects). Thanks!

**Comment by jaibyron** [comment]
> A fix for this has landed in the Dart SDK:
> 
>  breakpoints_red.mp4 
> And has also rolled into Flutter `master`:
> 
>  breakpoints_red_flutter.mp4 
> If you see unexpected breakpoint behaviour when using a version that includes these changes (which is currently only master/bleeding-edge builds, but will eventually be in a beta and then stable release), please file new issues with detailed descriptions of the behaviour you're seeing (and if possible, repro steps/projects). Thanks!

Hello @DanTup  - I'm currently seeing the behaviour outlined in the thread with unstable breakpoints.

I have the following setup:

Dart version 3.3.3
DevTools version 2.31.1
Flutter extension version 3.92.0

Which Dart version is this resolved in?

**Comment by DanTup** [comment]
> Which Dart version is this resolved in?

The fix is in the current Dart beta, but the Flutter beta looks a little behind so does not have it yet (but it will be in the next Flutter beta).

**Comment by DanTup** [comment]
The changes above are in the latest Flutter `beta`.

**Comment by felipecastrosales** [comment]
After this PR my debugger is strangely bad. When I set breakpoints and restart, I need to rebuild my app. It's been a terrible experience and I'm having to use debug print all the time.

**Comment by DanTup** [comment]
@felipecastrosales could you file a new issue with more information and steps to repro and please include whether it happens in a new project, or only this specific project? Thanks!

**Comment by felipecastrosales** [comment]
> @felipecastrosales could you file a new issue with more information and steps to repro and please include whether it happens in a new project, or only this specific project? Thanks!

I saw that you responded promptly, but at the moment I can't create an example.

But I reproduced it by simply placing a breakpoint and having it active, I hot-reloaded it; and my application got completely stuck; I needed to restart it. My app is complex and has integrations with native plugins and isolates; there may be interference with these points.

**Comment by ziarehmank** [comment]
> > @felipecastrosales could you file a new issue with more information and steps to repro and please include whether it happens in a new project, or only this specific project? Thanks!
> 
> I saw that you responded promptly, but at the moment I can't create an example.
> 
> But I reproduced it by simply placing a breakpoint and having it active, I hot-reloaded it; and my application got completely stuck; I needed to restart it. My app is complex and has integrations with native plugins and isolates; there may be interference with these points.

I have seen this issue many times, app gets stuck on hot reload, I never thought it could be related to debug points, but its happening often.
I am on flutter beta version.

 Flutter (Channel stable, 3.24.1, on macOS 14.6.1 23G93 darwin-arm64,

**Comment by felipecastrosales** [comment]
> I have seen this issue many times, app gets stuck on hot reload, I never thought it could be related to debug points, but its happening often. I am on flutter beta version.
> 
> Flutter (Channel stable, 3.24.1, on macOS 14.6.1 23G93 darwin-arm64,

Exactly, I hadn't realized it was due to that either, until I spoke to a teammate who said that when I added some the app was crashing... and coincidentally we were using the new version of Flutter.

But now, I realize that it is happening because of this... and that sometimes, to avoid having my reasoning interrupted, I end up putting debugPrint or using the dart developer log. :(

**Comment by ziarehmank** [comment]
Looks like i have to switch back to stable branch.

**Comment by DanTup** [comment]
Please file a new issue with details of this issue. Please test if you can reproduce in a sample app and if not, try to identify what's different between your app and the sample (it can be difficult to track down if it only happens in some projects but we have no information about what's in those projects).

> Looks like i have to switch back to stable branch.

If this issue only exists on master/beta, it's important we find and fix the issue before it goes out in the next stable, so any help in tracking it down is appreciated. Thanks!

**Comment by ziarehmank** [comment]
Yes the issue is in beta , I will try to create new app, and reproduce the issue.

**Comment by ziarehmank** [comment]
Today, when I started VSCode, the Dart plugin updated to version 3.96.0. I created a sample project, and the issue no longer occurs. Additionally, the issue is resolved in my other apps as well. It seems that the Dart plugin update may have fixed the problem.

**Comment by DanTup** [comment]
> Today, when I started VSCode, the Dart plugin updated to version 3.96.0. I created a sample project, and the issue no longer occurs. Additionally, the issue is resolved in my other apps as well. It seems that the Dart plugin update may have fixed the problem.

I can't think of anything in the extension that could affect this (this behaviour mostly comes from the debug adapter that lives in the SDK), so unless you also updated your SDK around the same time, I suspect the issue might just be intermittent.

If you see it again, please do open a new issue with the details (and confirm whether you can repro in a new project or only one specific project). I'm locking this issue to encourage new issues to be filed, because commenting about new issues here sends notifications to all of the people who have commented above. Thanks!
